Fix food spoiling on craft

ssorpg opened the pull request:
Hello everyone! New guy here again with another rather simple fix this time for the crafting spoilage problem. I hate having food go bad while I'm cooking it so, fingers crossed, this should fix that.

#### Summary
```SUMMARY: Bugfixes "Fix producing spoiled food when cooking non-spoiled ingredients."```

#### Purpose of change
Fixes #24580

#### Describe the solution
Adds the time it took to cook to the spoil timer of the finished product. `player_activity` class now has `int moves_used` that counts the moves used during a given craft. If we'd rather not stuff more things in there let me know and I'll figure out some other way. Additionally, I added an overload to `get_relative_rot` for use with crafting perishable items.

If anyone knows a better way to cast `time_duration` types to `double`, or the inverse, please let me know.

#### Describe alternatives you've considered
I initially attempted to update the spoilage timer continuously throughout the activity, but I found the code became a bit too confusing and messy for someone of my skill level to actually implement. To be honest, I'm not sure there's currently any meaningful difference between the two fixes. Perhaps some trigger that fires in between starting and finishing a craft that deals specifically with spoiled food. I can't think of one though.

In any case, I tried to add as little code as possible to fix it, so if someone more skilled than I implements a continuous function at some point, it should be rather easy to revert this change.

kevingranade commented:
What happens when crafting is intertupted?
Ideally it would catch up the age of the ingredients.

ssorpg commented:
So currently the way this works is it only modifies the age of the food product when you finish crafting. This means that if food spoils while you're crafting, it is actually spoiled in your inventory during the craft. However, once you finish crafting, the end product receives negative spoilage time equal to the time it took you to craft.

Bit of a long winded answer, but the end result is that yes, interrupting crafting will result in spoiled ingredients. I'll check right now just to make sure though.

ssorpg commented:
Okay after a little bit of testing:

Interrupting crafting after the food's expiration date means your ingredients become rotten. Once you've done this there's no going back - even resuming the task results in rotten cooked food.

kevingranade commented:
Thanks for checking, that sounds like it does exactly what we want.

kevingranade commented on src/item.cpp:
```diff
@@ -3096,6 +3096,10 @@ double item::get_relative_rot() const
 }
+double item::get_relative_rot( const double make_time ) const
+{
+    return to_turns<double>( rot ) / ( to_turns<double>( type->comestible->spoils ) + make_time );
```
Sorry I haven't had a chance to review this earlier.

I was testing this and going over the math, and I'm not sure it adds up.

item.rot is presumably somewhat updated at this point, so that's some rot value greater than what it was when the crafting started.

The relative rot this is retrieving is a number between 0.0 and 1.0 that tells us how far along in the spoilage of the food we are.
rot / type->comestible->spoils was already giving us a value that said that the input ingredient had aged during the craft time.
rot / ( spoils + make_time ) gives us something in between the spoilage when crafting started and the spoilage when it ends, but I'm not sure it's weighted far enough toward the start to be meaningful.

effectively this is like:
( pre-craft rot + craft time ) / ( total rot time + craft_time ) 
If craft time is large compared to total rot time, this is going to pull it closer to 1.0, which is rotten.

I think what you wanted to do is:
( pre-craft rot + craft time - craft time ) / total rot time
i.e.
( rot - craft time ) / total rot time

kevingranade replied:
You can calculate that number to subtract with ``get_rot_since( since, until, location )``

ssorpg commented:
Ah you're absolutely right! Not sure how I missed that computational bug. Doing a little bit of quick math, a food item with 99/100 (0.99) rot that then spends 10 turns crafting would turn into a product with 109/110 (0.9909) rot, which is greater than the rot it started with. This problem becomes more pronounced the longer you spend crafting. (20 turns crafting means it becomes 119/120 (0.9916) rot and so on)

I've modified it so now it should start with 99/100 rot and end with 99/100 rot. I also changed the `moves_used` variable to `start_turn` so we're not having to add to it every `do_turn()` and instead we simply get the turn we began crafting. In addition, I made sure that we get a new `start_time` if we stop and then resume crafting.

And thank you very much for the `get_rot_since()` function, I never would have found that!

nexusmrsep commented:
Jenkins wants src/crafting.cpp to be AStyled.

nexusmrsep commented:
@ssorpg if you can't achieve it in any other way you can always sneak in a query function to `item.h` like `time_point get_last_rot_check() const { return last_rot_check; }`.

ssorpg commented:
Okay, reworked it completely so now it fetches the exact items for the craft and their rot. Now I just apply their rot from when you started (or resumed) crafting at the end just before the craft is completed. I don't think there are any bugs in it but there's just an endless list of circumstances to test to determine if that's true.

Tatterdemalian commented:
I like this "only factor in spoil values when you start/resume crafting" option best. Crafting food would presumably include measures to keep flies away from the food while you're preparing it, picking off any manky bits (which, due to mold biology, doesn't actually reverse rotting, just prevents further rot), and so on.

kevingranade commented:
Went with #25675, this kind of grew some extra complexity so that ended up being a cleaner solution.
